api/src/wfl/module/batch.clj

```diff
@@ -63,10 +63,22 @@
 
 (defn update-workload!
   "Use transaction TX to batch-update WORKLOAD statuses."
+  [tx workload]
+  (letfn [(update! [{:keys [id] :as workload}]
+            (postgres/batch-update-workflow-statuses! tx workload)
+            (postgres/update-workload-status! tx workload)
+            (workloads/load-workload-for-id tx id))]
+    (util/when-> workload #(and (:started %) (not (:finished %))) update!)))
+
+(defn stop-workload!
+  "Use transaction TX to stop the WORKLOAD."
   [tx {:keys [id] :as workload}]
-  (if (or (:finished workload) (not (:started workload)))
-    workload
-    (do
-      (postgres/batch-update-workflow-statuses! tx workload)
-      (postgres/update-workload-status! tx workload)
-      (workloads/load-workload-for-id tx id))))
+  (letfn [(patch-workload! [cols]
+            (jdbc/update! tx :workload cols ["id = ?" id]))
+          (stop! [{:keys [id] :as workload}]
+            (let [now (OffsetDateTime/now)]
+              (when-not (:started workload)
+                (patch-workload! {:started now :finished now}))
+              (patch-workload! {:stopped now})
+              (workloads/load-workload-for-id tx id)))]
+    (util/unless-> workload #(or (:stopped %) (:finished %)) stop!)))
```

Comment:
because it isn't correct to set a workload to `:finished` if the workflows are currently processing - the update code won't run if the workload is `:finished`.